### ISSUE TITLE
Pickle does might not work with on_training_finished or on_epoch_finished

### ISSUE THREAD

**Issue by cancan101** [dnouri/nolearn#64]
Pickle does might not work with `on_training_finished` or `on_epoch_finished` if they are instance methods.

Related to #33

I am seeing this issue when following https://github.com/dnouri/nolearn/issues/32#issuecomment-73437819:
```
        on_epoch_finished=[
            early_stopping,
        ],
        on_training_finished=[early_stopping.load_best_weights]
```

**Comment by cortexml** [comment]
Not the best solution but it's a workaround for pickle
    
    class EarlyStopping(object):
    """
    Early stopping strategy which is used to prevent from overfitting the training data.
    """
    def __init__(self, patience = 100):
       self.patience = patience
        self.best_valid = np.inf
        self.best_valid_epoch = 0
        self.best_valid_accuracy = 0
        self.best_weights = None

    def __call__(self, nn, train_history):
        current_valid = train_history[-1]['valid_loss']
        current_epoch = train_history[-1]['epoch']
        current_valid_accuracy = train_history[-1]['valid_accuracy']

        # Use validation loss validation or accuracy 
        if current_valid_accuracy > self.best_valid_accuracy:
        #if current_valid < self.best_valid:
            self.best_valid = current_valid
            self.best_valid_epoch = current_epoch
            self.best_valid_accuracy = current_valid_accuracy
            self.best_weights = [w.get_value() for w in nn.get_all_params()]
        elif self.best_valid_epoch + self.patience < current_epoch:
            if nn.verbose:
                print("Early stopping.")
                print("Best valid loss was {:.6f} at epoch {} with accuracy {}.".format(
                    self.best_valid, self.best_valid_epoch, self.best_valid_accuracy))
            nn.load_weights_from(self.best_weights)
            if nn.verbose:
                print("Weights set.")
            raise StopIteration()

    class LoadBestWeight:
    def __init__(self, early_stopping):
        self.early_stopping = early_stopping

    def __call__(self, nn, train_history):
        print("Training stage finishes. Best valid loss was {:.6f} at epoch {} with accuracy {}.".format(
            self.early_stopping.best_valid, self.early_stopping.best_valid_epoch, \
                self.early_stopping.best_valid_accuracy))
        nn.load_weights_from(self.early_stopping.best_weights) 

usage: 

    on_epoch_finished=[early_stopping, ],
    on_training_finished=[LoadBestWeight(early_stopping)],

**Comment by dnouri** [comment]
Apparently there's a fix for pickling instance methods.  It's not pretty, but this code should do the trick:

```python
def _pickle_method(method):
    func_name = method.im_func.__name__
    obj = method.im_self
    cls = method.im_class
    return _unpickle_method, (func_name, obj, cls)


def _unpickle_method(func_name, obj, cls):
    for cls in cls.mro():
        try:
            func = cls.__dict__[func_name]
        except KeyError:
            pass
        else:
            break
    return func.__get__(obj, cls)


import copy_reg
import types

copy_reg.pickle(types.MethodType, _pickle_method, _unpickle_method)
```

Reference: first answer here http://stackoverflow.com/questions/1816958/cant-pickle-type-instancemethod-when-using-pythons-multiprocessing-pool-ma

Not sure if it's a good idea to include this in nolearn though.

**Comment by dnouri** [comment]
I like @cortexml's suggestion even better.  If anyone's up for adding an `EarlyStopping` class plus tests to nolearn, please go ahead.  I believe we can close this particular issue since it's a workaround for an implementation that isn't even part of the library.
